### PR TITLE
HIVE-26419: Use a different pool for DataNucleus' secondary connectio…

### PR DESCRIPTION
**Pull request by hsnusonic**
…n factory

### What changes were proposed in this pull request?

Introduce another connection pool for DataNucleus' secondary connection factory

### Why are the changes needed?

We now use the same connection pool for primary and secondary connection factory. It can possibly cause connection starvation when DataNucleus try to acquire a lock for schema validation or value generation but there is no available connection in the pool.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Running Impala test suites with pool size = 1 and all tests can pass

### CODE REVIEW

**Review by jfsii — APPROVED** [review]
It looks good to me - @deniskuzZ might be a good choice for a review/commit since they recently modified this area.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/PersistenceManagerProvider.java [hunk 1]
```diff
@@ -253,8 +253,12 @@ private static PersistenceManagerFactory initPMF(Configuration conf, boolean for
     } else {
       try {
         DataSource ds = (maxPoolSize > 0) ? dsp.create(conf, maxPoolSize) : dsp.create(conf);
+        // The secondary connection factory is used for schema generation, and for value generation operations.
+        // We should use a different pool for the secondary connection factory to avoid resource starvation.
+        // Since DataNucleus uses locks for schema generation and value generation, 2 connections should be sufficient.
+        DataSource ds2 = dsp.create(conf, /* maxPoolSize */ 2);
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/PersistenceManagerProvider.java:259]: According to the documentation:
Datastore connections are obtained from up to 2 connection factories. The primary connection factory is used for persistence operations, and optionally for value generation operations. The secondary connection factory is used for schema generation, and optionally for value generation operations.
Schema generation happens at HMS startup. 
Where do you see a potential issue with resource starvation?

Comment (hsnusonic) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/PersistenceManagerProvider.java:259]: Hi @deniskuzZ,
The issue is most easily to be observed with lots of add_partitions requests. When there is a write operation, DataNucleus will need to use ValueGenerator. Image there are 10 add_partitions requests, all have taken 1 connection from the pool. Now they are moving to value generation stage, there is a monitor lock in DataNucleus so only one will proceed and others are blocked. Even only one thread is trying to get connection from the pool, no connection is available (assuming we have 10 connections in the pool). This is the problem we use same pool for primary and secondary connection factory. Does it make sense to you?

**Review by deniskuzZ — APPROVED** [review]

**Review by deniskuzZ — CHANGES_REQUESTED** [review]
Please do not create separate ds2 for compaction (compactorPmf) it doesn't use ValueGenerator, just doing selects.

**Review by deniskuzZ — APPROVED** [review]
Tnx! LGTM +1

**Pull request merged**